File: freud/cli/analysis/telemetry/llm_cache.go
package telemetry

import (
	"encoding/json"
	"fmt"

	"github.com/atomitl/neopsyke/freud/cli/analysis"
)

// LLMCacheTelemetry aggregates LLM cache hit/miss/divergence stats from a JSONL events file.
func LLMCacheTelemetry(inputPath string) error {
	if !analysis.FileExists(inputPath) {
		fmt.Printf(`{"error": "Event log not found: %s"}`, inputPath)
		fmt.Println()
		return fmt.Errorf("event log not found: %s", inputPath)
	}

	eventTypes := map[string]bool{
		"llm_cache_hit":        true,
		"llm_cache_miss":       true,
		"llm_cache_divergence": true,
	}
	events, err := analysis.LoadJSONLMultiEvents(inputPath, eventTypes)
	if err != nil {
		return err
	}

	var hits, misses, divergences []map[string]interface{}
	for _, e := range events {
		switch analysis.GetString(e, "type") {
		case "llm_cache_hit":
			hits = append(hits, e)
		case "llm_cache_miss":
			misses = append(misses, e)
		case "llm_cache_divergence":
			divergences = append(divergences, e)
		}
	}

	totalCached := len(hits)
	totalReal := len(misses)
	totalCalls := totalCached + totalReal
	hitRate := 0.0
	if totalCalls > 0 {
		hitRate = float64(totalCached) / float64(totalCalls) * 100
	}

	var divergencePoint interface{}
	divergenceActor := ""
	divergenceCallSite := ""
	if len(divergences) > 0 {
		data := analysis.GetData(divergences[0])
		if v, ok := data["sequence_index"]; ok {
			divergencePoint = v
		}
		divergenceActor = analysis.GetString(data, "actor")
		divergenceCallSite = analysis.GetString(data, "call_site")
	}

	// Hits by actor
	actorHits := map[string]int{}
	for _, e := range hits {
		actor := analysis.GetString(analysis.GetData(e), "actor")
		if actor == "" {
			actor = "unknown"
		}
		actorHits[actor]++
	}

	// Hints
	var hints []string
	if divergencePoint != nil {
		hints = append(hints, fmt.Sprintf(
			"Divergence at seq %v (%s/%s): code change may have affected this call path.",
			divergencePoint, divergenceActor, divergenceCallSite))
	}
	if totalCalls > 0 && totalCached == 0 {
		hints = append(hints, "No cache hits: this may be a first run (record mode) or the cache file is empty.")
	}
	if totalCalls > 0 && hitRate == 100.0 {
		hints = append(hints, "All calls served from cache: fully deterministic replay.")
	}

	result := map[string]interface{}{
		"total_calls":          totalCalls,
		"cached_calls":         totalCached,
		"real_calls":           totalReal,
		"hit_rate_percent":     round2(hitRate),
		"divergence_count":     len(divergences),
		"divergence_point":     divergencePoint,
		"divergence_actor":     divergenceActor,
		"divergence_call_site": divergenceCallSite,
		"hits_by_actor":        actorHits,
		"hints":                hints,
	}

	data, _ := json.MarshalIndent(result, "", "  ")
	fmt.Println(string(data))
	return nil
}

func round2(f float64) float64 {
	return float64(int(f*100+0.5)) / 100
}

File: freud/legacy/py/telemetry/llm_cache.py
#!/usr/bin/env python3
"""LLM cache telemetry aggregator.

Parses JSONL events of types ``llm_cache_hit``, ``llm_cache_miss``, and
``llm_cache_divergence`` and outputs cache performance statistics as JSON.
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

from freud.py.common import repo_root


_CACHE_EVENT_TYPES = {"llm_cache_hit", "llm_cache_miss", "llm_cache_divergence"}


def _load_cache_events(path: Path) -> list[dict]:
    events = []
    for line in path.read_text(encoding="utf-8", errors="replace").splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            obj = json.loads(line)
        except json.JSONDecodeError:
            continue
        if obj.get("type") in _CACHE_EVENT_TYPES:
            events.append(obj)
    return events


def llm_cache_telemetry(input_path: str) -> None:
    path = Path(input_path)
    if not path.is_file():
        print(f'{{"error": "Event log not found: {input_path}"}}')
        sys.exit(1)

    events = _load_cache_events(path)

    hits = [e for e in events if e.get("type") == "llm_cache_hit"]
    misses = [e for e in events if e.get("type") == "llm_cache_miss"]
    divergences = [e for e in events if e.get("type") == "llm_cache_divergence"]

    total_cached = len(hits)
    total_real = len(misses)
    total_calls = total_cached + total_real
    hit_rate = (total_cached / total_calls * 100) if total_calls > 0 else 0.0

    divergence_point = None
    divergence_actor = None
    divergence_call_site = None
    if divergences:
        first = divergences[0]
        data = first.get("data", {})
        divergence_point = data.get("sequence_index")
        divergence_actor = data.get("actor", "")
        divergence_call_site = data.get("call_site", "")

    result = {
        "total_calls": total_calls,
        "cached_calls": total_cached,
        "real_calls": total_real,
        "hit_rate_percent": round(hit_rate, 2),
        "divergence_count": len(divergences),
        "divergence_point": divergence_point,
        "divergence_actor": divergence_actor,
        "divergence_call_site": divergence_call_site,
    }

    # Breakdown by actor
    actor_hits: dict[str, int] = {}
    for e in hits:
        actor = e.get("data", {}).get("actor", "unknown")
        actor_hits[actor] = actor_hits.get(actor, 0) + 1
    result["hits_by_actor"] = actor_hits

    # Tuning hints
    hints = []
    if divergence_point is not None:
        hints.append(
            f"Divergence at seq {divergence_point} ({divergence_actor}/{divergence_call_site}): "
            f"code change may have affected this call path."
        )
    if total_calls > 0 and total_cached == 0:
        hints.append("No cache hits: this may be a first run (record mode) or the cache file is empty.")
    if total_calls > 0 and hit_rate == 100.0:
        hints.append("All calls served from cache: fully deterministic replay.")
    result["hints"] = hints

    print(json.dumps(result, indent=2))


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Aggregate LLM cache telemetry from JSONL events.",
    )
    parser.add_argument(
        "events_jsonl",
        nargs="?",
        default=str(repo_root() / ".neopsyke/logs/latest-events.jsonl"),
        help="Path to events JSONL file",
    )
    args = parser.parse_args()

    llm_cache_telemetry(args.events_jsonl)


if __name__ == "__main__":
    main()
